File: Ahorn/entities/EntityWrappers/SolidModifier.jl
module VivHelperSolidModifier

using ..Ahorn, Maple

@mapdef Entity "VivHelper/SolidModifier" SolidModifier(x::Integer, y::Integer, width::Integer=8, height::Integer=8, Types::String="*Solid", EntitySelect::Bool=true)

const placements = Ahorn.PlacementDict(
    "~ Solid Modifier (Blank) (Viv's Helper)" => Ahorn.EntityPlacement(
        SolidModifier,
        "rectangle",
        Dict{String, Any}(
            "CornerBoostBlock" => 0,
            "TriggerOnBufferInput" => false,
            "TriggerOnTouch" => 0
        )
    ),
    "Corner Boost Block Modifier (Viv's Helper)" => Ahorn.EntityPlacement(
        SolidModifier,
        "rectangle",
        Dict{String,Any}(
            "CornerBoostBlock" => 0
        )
    ),
    "Touch Contact Modifier (Viv's Helper)" => Ahorn.EntityPlacement(
        SolidModifier,
        "rectangle",
        Dict{String,Any}(
            "TriggerOnBufferInput" => false,
            "TriggerOnTouch" => 0
        )
    )
)

Ahorn.editingOptions(entity::SolidModifier) = Dict{String, Any}(
    "Types" => String["*Celeste.Solid"],
    "CornerBoostBlock" => Dict{String, Any}(
        "Normal" => 0,
        "Corner Boost Block" => 1,
        "Retain Wall Speed" => 2
    ),
    "TriggerOnTouch" => Dict{String, Any}(
        "No Modifier" => 0,
        "On Touch" => 1,
        "On Touch + Bottom Contact" => 2
    ),
    "EntitySelect" => Dict{String, Any}(
        "All Valid Solids in Range" => true,
        "First Valid Solid in Range" => false
    )
)


Ahorn.resizable(entity::SolidModifier) = true, true
Ahorn.minimumSize(entity::SolidModifier) = 8,8

Ahorn.selection(entity::SolidModifier) = Ahorn.getEntityRectangle(entity)

sprite = "ahorn/VivHelper/solidModifierOutline.png"

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::SolidModifier, room::Maple.Room)
    for x in 0:(entity["width"]/8)-1, y in 0:(entity["height"]/8)-1
        Ahorn.drawSprite(ctx, sprite, 8*x + 4, 8*y + 4; tint=(0.5, 0.1625, 0.143, 0.5))
    end
end

end

File: Loenn/entities/EntityWrappers/SolidModifier.lua
local drawableRect = require('structs.drawable_rectangle')
local drawableText = require('structs.drawable_text')

return {{
    name = "VivHelper/SolidModifier",
    fieldInformation = {
        CornerBoostBlock = {fieldType = "integer", options = {
            ["Normal"] = 0,
            ["Climb at Any Speed"] = 1,
            ["Climb at Any Speed + Retain Wall Speed"] = 2
        }, editable = false},
        TriggerOnTouch = {fieldType = "integer", options = {
            ["No Modifier"] = 0,
            ["On Touch"] = 1,
            ["On Touch + Bottom Contact"] = 2
        }, editable = false},
    },
    fieldOrder = {"x","y","width","height","Types","EntitySelect","CornerBoostBlock","TriggerOnTouch","TriggerOnBufferInput"},--[[
        Remove legacy placements, retain editor functionality
    placements = {
        {name = "main", data = { width = 8, height = 8, 
            Types = "*Solid", EntitySelect = true,
            CornerBoostBlock = 0,  TriggerOnBufferInput = false, TriggerOnTouch = 0
        }},
        {name = "cb", data = { width = 8, height = 8, Types = "*Solid", EntitySelect = true, 
            CornerBoostBlock = 0
        }},
        {name = "touch", data = { width = 8, height = 8, Types = "*Solid", EntitySelect = true, 
            TriggerOnTouch = 0, TriggerOnBufferInput = false
        }}
    },]]--
    sprite = function(room,entity) return {
        drawableRect.fromRectangle("fill",entity.x,entity.y,entity.width,entity.height,{0.4,0.2,0.2,0.4}),
        drawableFunc.fromFunction(require('utils.drawing').printCenteredText, "~Legacy - SolidModifier [VivHelper]", entity.x,entity.y,entity.width,entity.height,nil,0.5)
    } end,
    depth = -100000
}, {
    name = "VivHelper/SolidModifier2",
    fieldInformation = {
        cornerBoostBlock = {fieldType = "integer", options = {
            {"Normal", 0},
            {"Based On Speed", -1},
            {"1px", 1}, {"2px", 2}, {"3px", 3}, {"4px", 4}, {"5px", 5}, {"6px", 6}, {"7px", 7}, {"8px", 8}, {"9px", 9}, {"10px", 10}, {"11px", 11}, {"12px", 12}, {"13px", 13}, {"14px", 14}, {"15px", 15}, {"16px", 16}, {"17px", 17}, {"18px", 18}, {"19px", 19}, {"20px", 20}, {"21px", 21}, {"22px", 22}, {"23px", 23}, {"24px", 24}, {"25px", 25}
        }, editable = true},
        TriggerOnTouch = {fieldType = "integer", options = {
            ["No Changes"] = 0,
            ["On Side Contact"] = 1,
            ["On Side + Bottom Contact"] = 2
        }, editable = false},
    },
    fieldOrder = {"x","y","width","height","Types","EntitySelect","CornerBoostBlock","RetainWallSpeed","TriggerOnTouch","TriggerOnBufferInput"},
    placements = {
        {name = "main", data = { width = 8, height = 8, 
            Types = "*Solid", EntitySelect = true,
            cornerBoostBlock = 0,  TriggerOnBufferInput = false, TriggerOnTouch = 0, RetainWallSpeed = false
        }},
        {name = "cb", data = { width = 8, height = 8, Types = "*Solid", EntitySelect = true, 
            cornerBoostBlock = -1, RetainWallSpeed = true
        }},
        {name = "touch", data = { width = 8, height = 8, Types = "*Solid", EntitySelect = true, 
            TriggerOnTouch = 0, TriggerOnBufferInput = false
        }}
    },
    sprite = function(room,entity) return {
        drawableRect.fromRectangle("fill",entity.x,entity.y,entity.width,entity.height,{0.4,0.2,0.2,0.4}),
        drawableText.fromText("SolidModifier [VivHelper]", entity.x,entity.y,entity.width,entity.height,nil,0.5)
    } end,
    depth = -100000
}}